### PR TITLE
Wip/hughsie/goodix tp

### PR DESCRIPTION
**Pull request by xulinkun**
Type of pull request:

- [X] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by xulinkun** [comment]
Hi, @hughsie ,here I go again... please review.

File: plugins/goodix-tp/fu-goodixtp-firmware.c [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+			   guint8 type,
+			   guint32 addr,
+			   const guint8 *buf,
+			   gint bufsz)
+{
+	FuGoodixtpFirmwarePrivate *priv = GET_PRIVATE(self);
+	if (bufsz > RAM_BUFFER_SIZE)
+		return;
+
+	memcpy(priv->fw_data + priv->fw_len, buf, bufsz);
```
Comment (hughsie) [inline plugins/goodix-tp/fu-goodixtp-firmware.c:79]: I'd rather a fu_memcpy_safe() here

File: plugins/goodix-tp/fu-goodixtp-brlb-firmware.c [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+		if (tmp_type == 0x0B || tmp_type == 0x01) {
+			g_debug("skip type[%x] subsystem", tmp_type);
+			continue;
+		}
+
+		for (gint j = 0; j < (gint)(tmp_size / RAM_BUFFER_SIZE); j++) {
+			fu_goodixtp_add_chunk_data(self,
+						   tmp_type,
+						   tmp_addr + offset,
+						   tmp_data + offset,
```
Comment (hughsie) [inline plugins/goodix-tp/fu-goodixtp-brlb-firmware.c:111]: this looks a bit risky; i think you need to check fw_offset+offset to see if it's in-range.

File: plugins/goodix-tp/fu-goodixtp-firmware.c [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-goodixtp-common.h"
+#include "fu-goodixtp-firmware.h"
+
+#define MAX_CHUNK_NUM 80
+
+struct FuGoodixChunkInfo {
```
Comment (hughsie) [inline plugins/goodix-tp/fu-goodixtp-firmware.c:16]: typedef struct pelase

File: plugins/goodix-tp/fu-goodixtp-firmware.c [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	guint32 flash_addr;
+};
+
+typedef struct {
+	FuFirmwareClass parent_instance;
+	guint32 version;
+	guint8 *fw_data;
+	gint fw_len;
+	gint index;
+	struct FuGoodixChunkInfo chunk_info[MAX_CHUNK_NUM];
```
Comment (hughsie) [inline plugins/goodix-tp/fu-goodixtp-firmware.c:27]: this seems like a big thing to put on the stack; can you make this a GPtrArray please.

**Comment by hughsie** [comment]
@xulinkun i've pushed another fixup patch; can you verify i've not broken anything please. There are a few more things I think we need to fix -- and then I think we're nearly complete.

**Comment by hughsie** [comment]
PVS Studio also picked up: 

    General Analysis	fu-goodixtp-hid-device.c:96	Medium	V1086	A call of the 'memcpy' function will lead to underflow of the buffer 'priv->patch_pid'.

**Comment by xulinkun** [comment]
I see, I'll submit again tomorrow. Google will use the fwupd instead of their own way on Chrome OS, so we need to hurry up.

**Comment by hughsie** [comment]
> Google will use the fwupd instead of their own way on Chrome OS

Yes, I hear the upstream maintainer is really strict about code quality :)

**Comment by xulinkun** [comment]
Hi, @hughsie , I've optimized goodix_firmware.c, please review it.

File: plugins/goodix-tp/fu-goodixtp-hid-device.c [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+#include <linux/input.h>
+#endif
+
+#include "fu-goodixtp-common.h"
+#include "fu-goodixtp-firmware.h"
+#include "fu-goodixtp-hid-device.h"
+
+typedef struct {
+	guint8 patch_pid[9];
+	guint8 patch_vid[4];
```
Comment (hughsie) [inline plugins/goodix-tp/fu-goodixtp-hid-device.c:22]: Question: what do we actually need the patch_pid and patch_vid _for_?

Comment (xulinkun) [reply plugins/goodix-tp/fu-goodixtp-hid-device.c:22]: Yes, patch_pid storage chip name such as GT7863, GT9966.... and vid storage firmware version. We need to print them.

**Comment by hughsie** [comment]
@xulinkun coverity was really upset with the meory safety in FuGoodixtpBrlbFirmware -- specifically the ownership of the data of the FuGoodixtpChunkInfo -- and I had very little idea of what the code was actually doing. I rewrote it using the standard primitives we have in fwupd -- which means the right thing should happen if you do `sudo fwupdtool firmware-parse gtx8-firmware.bin`

Can you check what I've done please, and verify the code still works on both hardware types. Feel free to add the missing data in the .rs file if that's helpful to you.

File: plugins/goodix-tp/fu-goodixtp.rs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    Berlinb,
+    Yellowstone,
+}
+
+#[derive(Parse)]
+struct GoodixBrlbHdr {
+    firmware_size: u32le,
+    checksum: u32le,
+    _unknown: [u8; 19],
+    vid: u16be,
```
Comment (hughsie) [inline plugins/goodix-tp/fu-goodixtp.rs:19]: @xulinkun is this definitely big endian? I thought all of brlb was LE?

**Comment by xulinkun** [comment]
Hi,@hughsie, thanks for your help. Writing flash must to align with 4K, I changed a little bit base your code.

File: plugins/goodix-tp/fu-goodixtp.rs (hunk 6 above)
Comment (xulinkun) [reply plugins/goodix-tp/fu-goodixtp.rs:19]: En... except vid is BE, the others is LE in brlb.

**Review by hughsie — APPROVED** [review]
LGTM.

**Pull request merged**